include/reader.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <algorithm>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <future>
+#include <memory>
+#include <stdexcept>
+#include <utility>
+
+#include <boost/interprocess/file_mapping.hpp>
+#include <boost/interprocess/mapped_region.hpp>
+
+class Reader {
+public:
+  virtual ~Reader() {}
+
+  virtual std::future<std::pair<const char*, size_t>> read(size_t len) = 0;
+};
+
+
+class FileReader: public Reader {
+public:
+  FileReader(const std::string& path, size_t blockSize):
+    File(open(path)), Cur(new char[blockSize]), Next(new char[blockSize])
+  {
+    std::setbuf(File, 0); // unbuffered, bitte
+  }
+
+  virtual ~FileReader() {
+    std::fclose(File);
+  }
+
+  virtual std::future<std::pair<const char*, size_t>> read(size_t len) {
+    std::swap(Cur, Next);
+    return std::async(
+      std::launch::async,
+      [](char* buf, size_t len, FILE* file) {
+        return std::pair<const char*, size_t>{
+          buf, std::fread(buf, 1, len, file)
+        };
+      },
+      Next.get(), len, File
+    );
+  }
+
+private:
+  static FILE* open(const std::string& path) {
+    if (path == "-") {
+      return stdin;
+    }
+    else {
+      FILE* f = std::fopen(path.c_str(), "rb");
+      if (!f) {
+        throw std::runtime_error(std::strerror(errno));
+      }
+      return f; 
+    }
+  }
+
+  FILE* File;
+  std::unique_ptr<char[]> Cur, Next;
+};
+
+namespace bip = boost::interprocess;
+
+class MemoryMappedFileReader: public Reader {
+public:
```

Comment:
likewise, the method implementations here should be defined in the .cpp.

I kinda' wonder whether we should have an enum with a value per reader type and then have a construction function declared here that could return the desired type with a base-class pointer. That way many of the headers would not be exposed.

Reply:
Done.